From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Paul Meyer <katexochen0@gmail.com>
Date: Mon, 1 Dec 2025 16:24:14 +0100
Subject: [PATCH] SmbiosMeasurementDxe: filter handoff table

We don't want to measure memory-size specific fields in the SMBIOS
handoff table, so the measurements stay constants between VMs of
different size. For this, we add the following things to the existing
filtering process:

- For SMBIOS table type 16, we filter the MaximumCapacity, as it seems
  to change between VMs of different memory size.
- For SMBIOS table type 17, we filter the Size field for the same
  reason.
- Tables of type 19 are removed completely for the measurement. That
  table contains the memory array mapped addresses, and most of its
  fields change depending on the memory presented by the VMM. Further,
  QEMU will add a second table of type 19 if the configured memory is
  larger than 4GiB. Therefore we cannot just filter the table content
  but remove the table as a whole.

The fields with the set filter will be zeroed by the FilterSmbiosEntry
function.

All relevant and mentioned structs are documented in
https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.2.0.pdf.

Co-authored-by: Moritz Sanft <58110325+msanft@users.noreply.github.com>
Signed-off-by: Paul Meyer <katexochen0@gmail.com>
---
 .../SmbiosMeasurementDxe.c                    | 19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

diff --git a/MdeModulePkg/Universal/SmbiosMeasurementDxe/SmbiosMeasurementDxe.c b/MdeModulePkg/Universal/SmbiosMeasurementDxe/SmbiosMeasurementDxe.c
index 9c7d3906be800dbd834feca9b36a7dc6e1230f33..3d80bfed2b3582f4e50940138b2eff2bd25ed1c9 100644
--- a/MdeModulePkg/Universal/SmbiosMeasurementDxe/SmbiosMeasurementDxe.c
+++ b/MdeModulePkg/Universal/SmbiosMeasurementDxe/SmbiosMeasurementDxe.c
@@ -68,10 +68,15 @@ SMBIOS_FILTER_TABLE  mSmbiosFilterType4BlackList[] = {
   { 0x04, OFFSET_OF (SMBIOS_TABLE_TYPE4, Voltage),           FIELD_SIZE_OF (SMBIOS_TABLE_TYPE4, Voltage),           0                                  },
   { 0x04, OFFSET_OF (SMBIOS_TABLE_TYPE4, CurrentSpeed),      FIELD_SIZE_OF (SMBIOS_TABLE_TYPE4, CurrentSpeed),      0                                  },
 };
+SMBIOS_FILTER_TABLE  mSmbiosFilterType16BlackList[] = {
+  { 0x10, OFFSET_OF (SMBIOS_TABLE_TYPE16, MaximumCapacity), FIELD_SIZE_OF (SMBIOS_TABLE_TYPE16, MaximumCapacity), 0 },
+};
 SMBIOS_FILTER_TABLE  mSmbiosFilterType17BlackList[] = {
   { 0x11, OFFSET_OF (SMBIOS_TABLE_TYPE17, SerialNumber), FIELD_SIZE_OF (SMBIOS_TABLE_TYPE17, SerialNumber), SMBIOS_FILTER_TABLE_FLAG_IS_STRING },
   { 0x11, OFFSET_OF (SMBIOS_TABLE_TYPE17, AssetTag),     FIELD_SIZE_OF (SMBIOS_TABLE_TYPE17, AssetTag),     SMBIOS_FILTER_TABLE_FLAG_IS_STRING },
   { 0x11, OFFSET_OF (SMBIOS_TABLE_TYPE17, PartNumber),   FIELD_SIZE_OF (SMBIOS_TABLE_TYPE17, PartNumber),   SMBIOS_FILTER_TABLE_FLAG_IS_STRING },
+  { 0x11, OFFSET_OF (SMBIOS_TABLE_TYPE17, Size),         FIELD_SIZE_OF (SMBIOS_TABLE_TYPE17, Size),         0                                  },
+  { 0x11, OFFSET_OF (SMBIOS_TABLE_TYPE17, ExtendedSize), FIELD_SIZE_OF (SMBIOS_TABLE_TYPE17, ExtendedSize), 0                                  },
 };
 SMBIOS_FILTER_TABLE  mSmbiosFilterType22BlackList[] = {
   { 0x16, OFFSET_OF (SMBIOS_TABLE_TYPE22, SerialNumber),        FIELD_SIZE_OF (SMBIOS_TABLE_TYPE22, SerialNumber),        SMBIOS_FILTER_TABLE_FLAG_IS_STRING },
@@ -97,6 +102,7 @@ SMBIOS_FILTER_STRUCT  mSmbiosFilterStandardTableBlackList[] = {
   { 0x04, mSmbiosFilterType4BlackList,  sizeof (mSmbiosFilterType4BlackList)/sizeof (mSmbiosFilterType4BlackList[0])   },
   { 0x0B, NULL,                         0                                                                              },
   { 0x0F, NULL,                         0                                                                              },
+  { 0x10, mSmbiosFilterType16BlackList, sizeof (mSmbiosFilterType16BlackList)/sizeof (mSmbiosFilterType16BlackList[0]) },
   { 0x11, mSmbiosFilterType17BlackList, sizeof (mSmbiosFilterType17BlackList)/sizeof (mSmbiosFilterType17BlackList[0]) },
   { 0x12, NULL,                         0                                                                              },
   { 0x16, mSmbiosFilterType22BlackList, sizeof (mSmbiosFilterType22BlackList)/sizeof (mSmbiosFilterType22BlackList[0]) },
@@ -449,7 +455,7 @@ GetSmbiosTableLength (
 VOID
 FilterSmbiosTable (
   IN OUT VOID  *TableAddress,
-  IN UINTN     TableLength
+  IN OUT UINTN *TableLength
   )
 {
   VOID   *TableAddressEnd;
@@ -457,13 +463,20 @@ FilterSmbiosTable (
   UINTN  TableEntryLength;
 
   TableEntry      = TableAddress;
-  TableAddressEnd = (VOID *)((UINTN)TableAddress + TableLength);
+  TableAddressEnd = (VOID *)((UINTN)TableAddress + *TableLength);
   while ((UINTN)TableEntry < (UINTN)TableAddressEnd) {
     TableEntryLength = GetSmbiosStructureSize (TableEntry, NULL);
     if (TableEntryLength == 0) {
       break;
     }
 
+    if (((SMBIOS_STRUCTURE *)TableEntry)->Type == SMBIOS_TYPE_MEMORY_ARRAY_MAPPED_ADDRESS) {
+      CopyMem (TableEntry, (VOID *)((UINTN)TableEntry + TableEntryLength), (UINTN)TableAddressEnd - (UINTN)TableEntry - TableEntryLength);
+      *TableLength -= TableEntryLength;
+      TableAddressEnd = (VOID *)((UINTN)TableAddress + *TableLength);
+      continue;
+    }
+
     FilterSmbiosEntry (TableEntry, TableEntryLength);
 
     TableEntry = (VOID *)((UINTN)TableEntry + TableEntryLength);
@@ -595,7 +608,7 @@ MeasureSmbiosTable (
       return;
     }
 
-    FilterSmbiosTable (TableAddress, TableLength);
+    FilterSmbiosTable (TableAddress, &TableLength);
 
     DEBUG ((DEBUG_INFO, "The final Smbios Table starts at: 0x%x\n", TableAddress));
     DEBUG ((DEBUG_INFO, "The final Smbios Table size: 0x%x\n", TableLength));
